### PR TITLE
chore: speed up kmean training for cosine

### PR DESCRIPTION
**Pull request by eddyxu**
`cargo bench --bench ivf_pq` resulted about 20% speed up over `cosine`

### CODE REVIEW

File: rust/lance-linalg/src/distance/cosine.rs [hunk 1]
```diff
@@ -125,6 +189,13 @@ fn cosine_scalar<T: Real + Sum>(x: &[T], x_norm: T, y: &[T]) -> T {
     T::one().sub(xy.div(x_norm.mul(y_sq.sqrt())))
 }
 
+#[inline]
+fn cosine_scalar_fast<T: Real + Sum>(x: &[T], x_norm: T, y: &[T], y_norm: T) -> T {
+    let xy = dot(x, y);
+    // 1 - xy / (sqrt(x_sq) * sqrt(y_sq))
+    T::one().sub(xy.div(x_norm.mul(y_norm)))
```
Comment (rok) [inline rust/lance-linalg/src/distance/cosine.rs:196]: Is there danger of overflow for `x_norm.mul(y_norm)` ?

**Review by rok — COMMENTED** [review]
Looks good, but I'm not familiar with this enough to comment. Just a question about overflow.

File: rust/lance-linalg/src/distance/cosine.rs (hunk 1 above)
Comment (eddyxu) [reply rust/lance-linalg/src/distance/cosine.rs:196]: That's a good question. we can probably cast it to f64 at the last step?

Comment (rok) [reply rust/lance-linalg/src/distance/cosine.rs:196]: What about this?
```suggestion
    T::one().sub(xy.div(x_norm).div(y_norm))
```

Comment (rok) [reply rust/lance-linalg/src/distance/cosine.rs:196]: Ah, that doesn't really help.

Comment (eddyxu) [reply rust/lance-linalg/src/distance/cosine.rs]: I upcast all of them to f64

**Review by wjones127 — APPROVED** [review]

**Pull request merged**